src/gc-debug.c

```diff
@@ -250,6 +253,120 @@ void gc_verify(jl_ptls_t ptls)
+    for (int region_i = 0; region_i < REGION_COUNT; region_i++) {
+        if (!regions[region_i].pages)
+            break;
+        region_t *region = &regions[region_i];
+        for (int pg_i = 0; pg_i <= region->ub; pg_i++) {
+            uint32_t line = region->allocmap[pg_i];
+            if (line) {
+                for (int j = 0; j < 32; j++) {
+                    if ((line >> j) & 1) {
+                        // for all pages in use
+                        jl_gc_pagemeta_t *pg = &region->meta[pg_i*32 + j];
+                        int p_n = pg->pool_n;
+                        int t_n = pg->thread_n;
+                        jl_ptls_t ptls2 = jl_all_tls_states[t_n];
+                        jl_gc_pool_t *p = &ptls2->heap.norm_pools[p_n];
+                        int osize = pg->osize;
+                        char *data = pg->data;
+                        char *page_begin = data + GC_PAGE_OFFSET;
+                        jl_taggedvalue_t *v = (jl_taggedvalue_t*)(data + GC_PAGE_OFFSET);
+                        char *lim = (char*)v + GC_PAGE_SZ - GC_PAGE_OFFSET - osize;
+                        // reset the freelist map to zero
+                        if (!freelist_zerod) {
+                            memset(freelist_map, 0, sizeof(freelist_map));
+                            freelist_zerod = 1;
+                        }
+                        // check for p in new newpages list
+                        jl_taggedvalue_t *halfpages = p->newpages;
+                        while (halfpages) {
+                            char *cur_page = gc_page_data((char*)halfpages - 1);
+                            if (cur_page == data) {
+                                lim = (char*)halfpages - 1;
+                                break;
+                            }
+                            halfpages = *(jl_taggedvalue_t**)cur_page;
+                        }
+                        // compute the freelist_map
+                        if (pg->nfree) {
+                            jl_taggedvalue_t *next = NULL;
+                            if (gc_page_data(p->freelist) == data) {
+                                // currently allocating on this page
+                                next = p->freelist;
+                                assert(page_metadata(next)->osize == osize);
+                                freelist_zerod = 0;
+                            }
+                            else if (pg->fl_begin_offset != (uint16_t)-1) {
+                                // part of free list exists on this page
+                                next = page_pfl_beg(pg);
+                                freelist_zerod = 0;
+                            }
+                            assert(halfpages || next);
+                            while (gc_page_data(next) == data) {
+                                int obj_id = (((char*)next) - page_begin) / osize;
+                                freelist_map[obj_id / 8] |= 1 << (obj_id % 7);
```

Comment:
Could also just use the page offset to calculate bit index (i.e. `/ sizeof(void*)` instead of `/ osize`) to make the code simpler. It's a fixed size array anyway....